src/modules/rlm_smtp/rlm_smtp.c

```diff
@@ -0,0 +1,614 @@
+	REQUEST			*request = mail_ctx->request;
+	VALUE_PAIR 		*vp;
+
+	fr_cursor_iter_by_da_init(&mail_ctx->cursor, &mail_ctx->request->packet->vps, attr_smtp_header);
+
+	/* Loop through all of the header elements and append them to a curl slist */
+	for ( vp = fr_cursor_current(&mail_ctx->cursor); vp; vp = fr_cursor_next(&mail_ctx->cursor)) {
+		if (strncmp(vp->vp_strvalue, date, 5) == 0) {
+			RDEBUG2("date set manually to: %s", vp->vp_strvalue);
+			mail_ctx->time_status = LOADED;
+		}
+		RDEBUG2("Adding Header: %s", vp->vp_strvalue);
+		mail_ctx->header = curl_slist_append(mail_ctx->header, vp->vp_strvalue);
+	}
+	/* If no header elements could be found, there is an error */
+	if ( mail_ctx->header == NULL) {
+		RDEBUG2("Header elements could not be added");
+ 		return -1;
+	}
+	/* If no time stamp was specified, add the time that the modules initially received the request */
+	if (mail_ctx->time_status == UNLOADED){
+		time_out = FR_SBUFF_OUT(mail_ctx->time_str, sizeof(mail_ctx->time_str));
+		RDEBUG2("No date was provided, setting automatically");
+		fr_time_strftime_local(&time_out, fr_time(), "DATE: %a, %d %b %Y %T %z, (%Z) \r\n");
+		RDEBUG2("Adding Header: %s", mail_ctx->time_str);
+		mail_ctx->header = curl_slist_append(mail_ctx->header, mail_ctx->time_str);
+	}
+	RDEBUG2("Finished generating the curl_slist for the header elements");
+	return 0;
+}
+
+/*
+ * Add the Body elements to the email
+ */
+static size_t body_source(char *ptr, size_t size, size_t nmemb, void *uctx)
+{
+	fr_mail_ctx 		*mail_ctx = uctx;
+	fr_dbuff_t		out;
+	REQUEST			*request = mail_ctx->request;
+	VALUE_PAIR 		*vp;
+
+	fr_dbuff_init(&out, (uint8_t *)ptr, (size * nmemb));  /* Wrap the output buffer so we can track our position easily */
+	vp = fr_cursor_current(&mail_ctx->body_cursor);
+
+	/* Copy the vp into the email. If it cannot all be loaded, return the amount of memory that was loaded and get called again */
+	if (fr_dbuff_memcpy_in_partial(&out, &mail_ctx->vp_in, SIZE_MAX) < fr_dbuff_remaining(&mail_ctx->vp_in)) {
+		RDEBUG2("%zu bytes used (partial copy)", fr_dbuff_used(&out));
+		return fr_dbuff_used(&out);
+	}
+	/* Once this value pair is fully copied, prepare for the next element */
+	vp = fr_cursor_next(&mail_ctx->body_cursor);
+	if (vp) {
+		fr_dbuff_init(&mail_ctx->vp_in, (uint8_t const *)vp->vp_strvalue, vp->vp_length);
+
+	}
+	RDEBUG2("%zu bytes used (full copy)", fr_dbuff_used(&out));
+	return fr_dbuff_used(&out);
+}
+
+/*
+ * Adds every SMTP_Attachments file to the email as a MIME part
+ */
+static int attachments_source(fr_mail_ctx *uctx, curl_mime *mime, rlm_smtp_t const *inst)
+{
+	curl_mimepart		*part;
+	REQUEST			*request = uctx->request;
+	VALUE_PAIR 		*vp;
+	int 			attachments_set = 0;
+
+	vp = fr_cursor_iter_by_da_init(&uctx->cursor, &uctx->request->packet->vps, attr_smtp_attachment_file);
+
+	/* Initialize the buffer to write the file path */
+	fr_sbuff_init_talloc(uctx, &uctx->path_buffer, &uctx->sbuff_ctx, 0, SIZE_MAX);
```

Comment:
You're leaking (ish, it will be freed eventually) memory here every time attachment_source is called, you should call this once when the uctx is created, then set a marker so you can just copy over the trailing portion of the path.  Also means you don't need to run the logic to figure out if there's a trailing slash each time.

As I mentioned in slack, you should drop a marker after copying in the template_dir string (and the trailing slash), and reset the position each time you want to process a new vp.